src/hb-ot-layout-common.hh

```diff
@@ -1574,15 +1574,25 @@ static void ClassDef_remap_and_serialize (hb_serialize_context_t *c,
                                           const hb_set_t &glyphset,
                                           const hb_map_t &gid_klass_map,
                                           hb_sorted_vector_t<HBGlyphID> glyphs,
-                                          hb_sorted_vector_t<unsigned> klasses,
+                                          const hb_set_t &klasses,
                                           hb_map_t *klass_map /*INOUT*/)
 {
-  bool has_no_match = glyphset.get_population () > gid_klass_map.get_population ();
+  if (!klass_map)
+  {
+    auto it =
+    + glyphs.iter ()
+    | hb_map_retains_sorting ([&] (const HBGlyphID& gid) -> hb_pair_t<hb_codepoint_t, unsigned>
+                              {
+                                return hb_pair ((hb_codepoint_t)gid, gid_klass_map[gid]);
+                              })
+    ;
+    ClassDef_serialize (c, it);
+    return;
+  }
 
-  hb_map_t m;
-  if (!klass_map) klass_map = &m;
+  if (glyphset.get_population () > gid_klass_map.get_population ())
```

Comment:
Would it make more sense to check for the presence of a mapping for key 0 if present in glyphset and then add if it's missing?

Reply:
There's case that if font file has 4 glyphs which are explicitly assigned to class 1,2,3,4 respectively, then we need a mapping like 1->0, 2->1, 3->2, 4->3.  
In this case 0->0 mapping is not needed.

Reply:
oh we have a map from gid->class, but not have a class->gid map here

Reply:
Can you add a comment here explaining the purpose of the check.

Reply:
I think you're right we could check for the presence of a class 0 mapping. Given we only have gid->class mapping, we need to traverse the whole map or build a reverse map to check this. 

Another reason is to be consistent with fonttools:
https://github.com/fonttools/fonttools/blob/master/Lib/fontTools/subset/__init__.py#L436

Reply:
Ok, reviewed again and the current approach seems fine. No need to change it but can you add a comment though that explains what the check is for.

Reply:
comments added